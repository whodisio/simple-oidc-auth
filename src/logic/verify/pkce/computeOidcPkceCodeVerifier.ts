import { sha256 } from 'cross-sha256';

/**
 *
 * requirements
 * - sufficient entropy - âœ…, challenge uuid = a uuid, completely random -> increased to 64 char through sha256
 *
 * ref
 * - https://datatracker.ietf.org/doc/html/rfc7636#section-4.1
 */
export const computeOidcPkceCodeVerifier = ({
  oidcRequestUuid,
  userSessionUuid,
}: {
  /**
   * the primary key identifier of the request, persisted by your backend
   */
  oidcRequestUuid: string;

  /**
   * the primary key identifier of the current user's session, if any
   */
  userSessionUuid: string | null;
}): string =>
  [
    'opcv',
    new sha256()
      .update(
        JSON.stringify({
          purpose: 'oidcPkceCodeVerifier',
          oidcRequestUuid,
          userSessionUuid,
        }),
      )
      .digest('hex'),
  ].join('_');
