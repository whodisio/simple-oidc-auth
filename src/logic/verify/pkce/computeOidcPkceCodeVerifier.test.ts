import { uuid } from '../../../deps';
import { computeOidcPkceCodeVerifier } from './computeOidcPkceCodeVerifier';

describe('computeOidcPkceCodeVerifier', () => {
  it('should be deterministic', () => {
    const inputs = {
      oidcRequestUuid: uuid(),
      userSessionUuid: null,
    };
    expect(computeOidcPkceCodeVerifier(inputs)).toEqual(
      computeOidcPkceCodeVerifier(inputs),
    );
  });
  it('should look nice', () => {
    const inputs = {
      oidcRequestUuid: '0fb30413-49ab-483d-8b8e-19e3d3e96eec',
      userSessionUuid: null,
    };
    expect(computeOidcPkceCodeVerifier(inputs)).toMatch(/^opcv_/); // starts with opcv
    expect(computeOidcPkceCodeVerifier(inputs)).toMatchSnapshot();
  });
});
