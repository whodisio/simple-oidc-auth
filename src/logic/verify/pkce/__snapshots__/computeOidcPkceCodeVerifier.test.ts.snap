// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeOidcPkceCodeVerifier should look nice 1`] = `"opcv_b937166d9627c0c1289d0fae73074b7a31d796d076e9adb843738f059ecd485c"`;
